packages/test/test-end-to-end-tests/src/test/idCompressor.spec.ts

```diff
@@ -0,0 +1,326 @@
+		sharedMap1.set(firstDecompressedId, "value1");
+
+		[secondId, thirdId].forEach((id, index) => {
+			assert(getIdCompressor(sharedMap2) !== undefined);
+			const decompressedId = getIdCompressor(sharedMap2).decompress(id);
+			decompressedIds.push(decompressedId);
+			sharedMap2.set(decompressedId, `value${index + 2}`);
+		});
+
+		// should be negative
+		assert(getIdCompressor(sharedMap1).normalizeToOpSpace(firstId) < 0);
+		assert(getIdCompressor(sharedMap2).normalizeToOpSpace(secondId) < 0);
+		assert(getIdCompressor(sharedMap2).normalizeToOpSpace(thirdId) < 0);
+
+		await provider.ensureSynchronized();
+
+		assert(getIdCompressor(sharedMap1).normalizeToOpSpace(firstId) > 0);
+		assert(getIdCompressor(sharedMap2).normalizeToOpSpace(secondId) > 0);
+		assert(getIdCompressor(sharedMap2).normalizeToOpSpace(thirdId) > 0);
+
+		decompressedIds.forEach((id, index) => {
+			assert.equal(sharedMap1.get(id), `value${index + 1}`);
+			assert.equal(sharedMap2.get(id), `value${index + 1}`);
+		});
+	});
+
+	// IdCompressor is at container runtime level, which means that individual DDSs
+	// in the same container should have the same underlying compressor state
+	it("container with multiple DDSs has same compressor state", async () => {
+		assert(getIdCompressor(sharedMap1) !== undefined);
+		assert(getIdCompressor(sharedCell1) !== undefined);
+
+		// 2 IDs in the map compressor, 1 in the cell compressor
+		// should result in a local count of 3 IDs
+		const sharedMapCompressedId = getIdCompressor(sharedMap1).generateCompressedId();
+		const sharedMapDecompressedId =
+			getIdCompressor(sharedMap1).decompress(sharedMapCompressedId);
+		const sharedMapCompressedId2 = getIdCompressor(sharedMap1).generateCompressedId();
+		const sharedMapDecompressedId2 =
+			getIdCompressor(sharedMap1).decompress(sharedMapCompressedId2);
+		const sharedCellCompressedId = getIdCompressor(sharedCell1).generateCompressedId();
+		const sharedCellDecompressedId =
+			getIdCompressor(sharedMap1).decompress(sharedCellCompressedId);
+
+		// Generate an op so the idCompressor state is actually synchronized
+		// across clients
+		sharedMap1.set(sharedMapDecompressedId, "value");
+
+		assert.strictEqual(
+			(getIdCompressor(sharedMap1) as any).localIdCount,
+			(getIdCompressor(sharedCell1) as any).localIdCount,
+		);
+
+		await provider.ensureSynchronized();
+
+		assert.strictEqual(
+			getIdCompressor(sharedMap1).recompress(sharedMapDecompressedId),
+			getIdCompressor(sharedCell1).recompress(sharedMapDecompressedId),
+		);
+
+		assert.strictEqual(
+			getIdCompressor(sharedMap1).recompress(sharedMapDecompressedId2),
+			getIdCompressor(sharedCell1).recompress(sharedMapDecompressedId2),
+		);
+
+		assert.strictEqual(
+			getIdCompressor(sharedMap1).recompress(sharedCellDecompressedId),
+			getIdCompressor(sharedCell1).recompress(sharedCellDecompressedId),
+		);
+
+		assert.strictEqual(sharedMap1.get(sharedMapDecompressedId), "value");
+	});
+});
```

Comment:
There should be tests where compressed ids are part of ops and summaries and then validations that the ids can are correct / consitent. Few example tests:
1. The same id is negative in one client (before finalizing range) and is positive in other. Functionalities built on top of these ids work consistently.
2. A summarizer summarizes the ids in idCompressor. A new container loads from this summary and can correctly regenerate the ids.
3. Compressed ids in summarizer client cannnot be -ve.
4. Tests that validate op resubmit, i.e., compressed ids in disconnected state are correctly resubmitted when connected.
  4.1. Multiple compressed ids sent in disconnected state are correct resubmitted.
4. Compressed ids generated when container is detached should work correctly when container attaches.
6.

Reply:
I believe I have coverage of these scenarios now. This was a great list for helping find a few issues! There's still an issue with sending the stashed ops state over the wire and failing the PendingStateManager content equality check that I'm looking into, but should be ready for another round of review I think.